raphtory/src/algorithms/k_core.rs

```diff
@@ -0,0 +1,154 @@
+use crate::{
+    core::{entities::VID, state::compute_state::ComputeStateVec},
+    db::{
+        api::view::{GraphViewOps, VertexViewOps},
+        task::{
+            context::Context,
+            task::{ATask, Job, Step},
+            task_runner::TaskRunner,
+            vertex::eval_vertex::EvalVertexView,
+        },
+    },
+};
+use std::{
+    cmp,
+    collections::{HashMap, HashSet},
+};
+
+#[derive(Clone, Debug)]
+struct KCoreState {
+    alive: bool,
+}
+
+impl KCoreState {
+    fn new() -> Self {
+        Self { alive: true }
+    }
+}
+
+/// Determines which nodes are in the k-core for a given value of k
+///
+/// # Arguments
+///
+/// * `g` - A reference to the graph
+/// * `k` - Value of k such that the returned vertices have degree > k (recursively)
+/// * `iter_count` - The number of iterations to run
+/// * `threads` - number of threads to run on
+///
+/// # Returns
+///
+/// A hash set of vertices in the k core
+///
+pub fn k_core_set<G>(graph: &G, k: usize, iter_count: usize, threads: Option<usize>) -> HashSet<VID>
+where
+    G: GraphViewOps,
+{
+    let ctx: Context<G, ComputeStateVec> = graph.into();
+
+    let step1 = ATask::new(move |vv| {
+        let deg = vv.degree();
+        let state: &mut KCoreState = vv.get_mut();
+        state.alive = if deg < k { false } else { true };
+        Step::Continue
+    });
+
+    let step2 = ATask::new(
+        move |vv: &mut EvalVertexView<'_, G, ComputeStateVec, KCoreState>| {
+            let prev: bool = vv.prev().alive;
+            let current = vv
+                .neighbours()
+                .into_iter()
+                .map(|n| n.prev().alive)
+                .filter(|b| *b)
+                .count()
+                >= k;
+            let state: &mut KCoreState = vv.get_mut();
+            if current != prev {
+                state.alive = current;
+                Step::Continue
+            } else {
+                Step::Done
+            }
+        },
+    );
+
+    let mut runner: TaskRunner<G, _> = TaskRunner::new(ctx);
+
+    runner
+        .run(
+            vec![Job::new(step1)],
+            vec![Job::read_only(step2)],
+            KCoreState::new(),
+            |_, _, _, local| {
+                local
+                    .iter()
+                    .enumerate()
+                    .map(|(v_ref, state)| (v_ref.into(), state.alive))
+                    .collect::<HashMap<_, _>>()
+            },
+            threads,
+            iter_count,
+            None,
+            None,
+        )
+        .into_iter()
+        .filter(|(k, v)| *v)
+        .map(|(k, v)| k)
+        .collect()
```

Comment:
you don't need to make a temporary `HashMap` and filter on it again. Just do the filter in the closure directly and return the set.

Reply:
Ahh of course 🤦‍♀️ thanks got it!